File: No_865/Smallest Subtree with all the Deepest Nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* subtreeWithAllDeepest(TreeNode* root) {
        int leftHeight = findHeight(root->left);
        int rightHeight = findHeight(root->right);
        if(leftHeight == rightHeight) {
            return root;
        } else if(leftHeight > rightHeight) {
            return subtreeWithAllDeepest(root->left);
        }
        return subtreeWithAllDeepest(root->right);
    }
    
    int findHeight(TreeNode *root) {
        if(root == nullptr) return 0;
        return max(findHeight(root->left), findHeight(root->right)) + 1;
    }
    
   
};

File: No_865/Smallest Subtree with all the Deepest Nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def subtreeWithAllDeepest(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def dfs(node: Optional[TreeNode]) -> tuple[int, Optional[TreeNode]]:
            if not node:
                return 0, None

            left_depth, left_node = dfs(node.left)
            right_depth, right_node = dfs(node.right)

            if left_depth == right_depth:
                return left_depth + 1, node
            elif left_depth > right_depth:
                return left_depth + 1, left_node
            else:
                return right_depth + 1, right_node

        _, result = dfs(root)
        return result